shell/platform/android/io/flutter/embedding/android/FlutterView.java

```diff
@@ -508,30 +509,50 @@ private int guessBottomKeyboardInset(WindowInsets insets) {
   public final WindowInsets onApplyWindowInsets(@NonNull WindowInsets insets) {
     WindowInsets newInsets = super.onApplyWindowInsets(insets);
 
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+      Insets systemGestureInsets = insets.getSystemGestureInsets();
+      viewportMetrics.systemGestureInsetTop = systemGestureInsets.top;
+      viewportMetrics.systemGestureInsetRight = systemGestureInsets.right;
+      viewportMetrics.systemGestureInsetBottom = systemGestureInsets.bottom;
+      viewportMetrics.systemGestureInsetLeft = systemGestureInsets.left;
+    }
+
     boolean statusBarVisible = (SYSTEM_UI_FLAG_FULLSCREEN & getWindowSystemUiVisibility()) == 0;
     boolean navigationBarVisible =
         (SYSTEM_UI_FLAG_HIDE_NAVIGATION & getWindowSystemUiVisibility()) == 0;
 
     if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-      int mask = 0;
+      int mask = android.view.WindowInsets.Type.ime();
       if (navigationBarVisible) {
         mask = mask | android.view.WindowInsets.Type.navigationBars();
       }
       if (statusBarVisible) {
         mask = mask | android.view.WindowInsets.Type.statusBars();
       }
-      mask = mask | android.view.WindowInsets.Type.ime();
+      Insets uiInsets = insets.getInsets(mask);
 
-      Insets finalInsets = insets.getInsets(mask);
-      viewportMetrics.paddingTop = finalInsets.top;
-      viewportMetrics.paddingRight = finalInsets.right;
+      viewportMetrics.paddingTop = uiInsets.top;
+      viewportMetrics.paddingRight = uiInsets.right;
       viewportMetrics.paddingBottom = 0;
-      viewportMetrics.paddingLeft = finalInsets.left;
+      viewportMetrics.paddingLeft = uiInsets.left;
 
       viewportMetrics.viewInsetTop = 0;
       viewportMetrics.viewInsetRight = 0;
-      viewportMetrics.viewInsetBottom = finalInsets.bottom;
+      viewportMetrics.viewInsetBottom = uiInsets.bottom;
       viewportMetrics.viewInsetLeft = 0;
+
+      // TODO(garyq): Expose the full rects of the display cutout.
+
+      // Take the max of the display cutout insets and existing insets to merge them
+      DisplayCutout cutout = insets.getCutout();
+      if (cutout != null) {
+        Insets waterfallInsets = cutout.getWaterfallInsets();
+        viewportMetrics.systemGestureInsetTop = Math.max(Math.max(viewportMetrics.systemGestureInsetTop, waterfallInsets.top), cutout.getSafeInsetTop());
```

Comment:
how did you decide it is a systemGestureInset vs a padding? Both seem potentially ok but if you use systemGestureInset, you need to change the media query doc since it seems really specific to solve one problem currentyl.

Reply:
Ahh upon further review of what each is meant to be, padding explicitly mentions notches and cutouts, so I'll proceed with padding.